### DISCUSSION TITLE
First steps

### DISCUSSION THREAD

**Discussion by kimmolinna** [kimmolinna/pyfingrid#2] [General]
@masipila This is a proof of concept start for pyfingrid. You can use the Selenium library to login and to get authorisation token which you can use with the requests library.

**Comment by masipila** [comment]
Heya!

Thumbs up! Here are the first observations...

I'm running Windows 10 and Python 3.10. I normally use Git Bash for Windows as my terminal, well, because I don't like the Windows command prompt. 

I cloned the repo to localhost, figured out that I have to 

```
pip install selenium
pip install webdriver_manager
```
After that, I was able to run pyfingrid.py using the Git Bash terminal. 

- Selenium opened Chrome as expected and I logged in to Datahub using strong authentication. 
- After I had logged in to Datahub and "pressed any key", I saw nothing in the terminal except one deprecation warning, but concluded that it's unrelated since it's only a deprecation warning 
"pyfingrid.py:5: DeprecationWarning: executable_path has been deprecated, please pass in a Service object"

I then decided to try to use the native Windows command prompt. There I got the agreement data printed in the terminal (after a bunch of errors / warnings:
```

pyfingrid.py:5: DeprecationWarning: executable_path has been deprecated, please pass in a Service object
  driver = webdriver.Chrome(ChromeDriverManager().install())
[16300:24276:0111/220458.525:ERROR:google_update_settings.cc(236)] Failed opening key Software\Google\Update\ClientStateMedium\{8A69D345-D564-463c-AFF1-A69D9E530F96} to set usagestats; result: 5

DevTools listening on ws://127.0.0.1:64560/devtools/browser/f4fc324c-9d0f-4652-9256-1674e85f50d1
Press any key to continue after authentication[16300:23076:0111/220501.988:ERROR:device_event_log_impl.cc(215)] [22:05:01.988] USB: usb_service_win.cc:415 Could not read device interface GUIDs: The system cannot find the file specified. (0x2)
[16300:23076:0111/220501.995:ERROR:device_event_log_impl.cc(215)] [22:05:01.995] USB: usb_device_handle_win.cc:1045 Failed to read descriptor from node connection: A device attached to the system is not functioning. (0x1F)
[16300:23076:0111/220502.007:ERROR:device_event_log_impl.cc(215)] [22:05:02.008] USB: usb_device_handle_win.cc:1045 Failed to read descriptor from node connection: A device attached to the system is not functioning. (0x1F)
[16300:23076:0111/220502.009:ERROR:device_event_log_impl.cc(215)] [22:05:02.009] USB: usb_device_handle_win.cc:1045 Failed to read descriptor from node connection: A device attached to the system is not functioning. (0x1F)
[16300:16948:0111/220558.663:ERROR:util.cc(133)] Can't create base directory: C:\Program Files\Google\GoogleUpdater

b'[{agreement-content-in-json-format-here}]'
```

I'm eventually planning to run this on my Raspberry Pi where I don't have a GUI operating system. For this use case I would personally prefer that we would have two scripts:

- One script (this one which we have now), which I could run on my Windows laptop. It would launch Chrome for the authentication and print the session token
- A second script, which I would run on my Raspberry. Obviously I would need to set the session token for example as an environment variable on my Raspberry so that this second script could then make all the subsequent API calls using the token.

**Reply by kimmolinna** [reply]
@masipila Thank you very much. I just tested with Mac. I have to change use of webdriver. Luckily I have some ideas how to fix it.

**Comment by kimmolinna** [comment]
@masipila Now I have tested the latest version with Windows 10/Python 3.10, Ubuntu 22.04.1 LTS/Python 3.10 and Apple M1/Ventura 13.1 and it should work. I'm not sure should I put cookies to session, but at the moment I do that. Now the script just prints metering points to be sure it's working.

**Reply by masipila** [reply]
The latest version prints the expected output for me, too.

I know that this is heavily work in progress, but two nits and one real comment.

1. fg.get_cookies() now returns a key-value dictionary of cookie variables. 
However, your docblock says that it would be returning the token as a string. 

2. I find the name of the fg.get_session() method is a bit confusing. I propose that we rename this method as fg.prepare_session and comment it something like _Prepares and returns a Requests Session object so that the cookie variables are correctly populated for subsequent HTTP requests._

3. What concerns me here slightly is that the fact that we need to pass the whole Session object (with all cookie variables that we got from Selenium) in order to make the actual request to read the metering points. We need to eventually be able to de-couple the Selenium part from the subsequent requests so that we can run the API requests independently without the need to do the strong authentication every time we run the script. Looking at the cookie variables with the browser's web developer tools, Datahub seems to update the value of the XSRF-token on every page load to prevent cross-site request forgery. I'm not sure if that is relevant to make the calls against the /_api/ endpoints, though. In other words, I wonder if the subsequent calls will work with the original cookie variable values that we sniffed with Selenium...

Cheers,
Markus

**Reply by kimmolinna** [reply]
A couple of comments:
1. Thanks. This is now fixed. Actually I deleted the whole `fg.get_token`-function because I save authorisation-header to the session directly.
2. `fg.get_session` returns a new session with the proper cookies. And you use this session in all function calls after that. It's similar to the login-function in pycaruna.
3. I think this is the easiest way to handle the cookies and tokens. Of course you could pass the cookies to every call and a create a new session everytime but I think It's not wise. The whole Selenium part is possible to delete if it's possible to create a session with a necessary authorisation directly. At the moment the Selenium driver is deleted when the cookies are saved.

**Reply by masipila** [reply]
Regarding the Selenium part and establishing the session for the API calls that fetch the actual data.

The fundamental difference between Caruna Plus and Fingrid Datahub is that Datahub requires a strong authentication. Which means that in pycaruna we pass the username and password to the script as input, it logs in on behalf of us without needing a human interaction, fetches the data and then we save it wherever we want.

With Datahub, there is the strong authentication part, which makes this conceptually difficult, if / when the target is to have a daily scheduled job that fetches the consumption data of the previous day. Even if we would be able to pass the bank's username and password for Selenium (it's conceptually doable and automatable, but I would never ever store those credentials anywhere), there's the second factor of multi-factor authentication which is typically the mobile app of your bank.  So the benefit of scripting the data fetching is dramatically decreasing, because it needs a human interaction.   

What I dream of here with pyfingrid is something like this:

1. We have one script, let it be datahub-authentication.py. This script can use Selenium and require human interaction exactly like it does now. The sole purpose of this script is that we 
a) are able to establish the session that is required to make subsequent API calls
b) output the session in a format which we can persist, for example a Python dictionary with key-value pairs like now.

2. We then have another script, which fetches the data. Let this be datahub-fetch.py. This script should be able to be executed without a human interaction. We would pass the session components which we persisted earlier to this one.

I think we are already super close of this, because fingrid.py has methods that support both of these. So I could write the first script (datahub-authentication.py) and call fg.get_session() from it, and persist the dictionary how I see best.

Then, I could already now write the datahub-fetch.py, which would first read the session components from where I persisted it, create the Requests Session object and then make the API calls that fetch the data.

**Reply by kimmolinna** [reply]
This should be possible with `pyfingrid` already now. `fg.get_cookies()` is basically `datahub-authentication.py` and `fg.get_session(cookies)` returns you a session with the cookies and the authentication header. If you want to transfer the cookies between the operating systems that should be quite trivial.

**Reply by masipila** [reply]
Yep...

I was able to make a request that fetched the consumption data from the /_api/GetConsumptionData endpoint. I'll clean up the code slightly and will share it here soon with sample usage.

After that it's a question of figuring out how soon will the once established session time out from the server side. And then figuring out a mechanism to keep the session alive.

**Reply by masipila** [reply]
I'm currently running a script that calls the getConsumptionData endpoint once per hour. Let's see if that is enough to keep the session alive and if that is frequent enough, then I'll try to decrease the frequency.

And then I need to learn how to create a pull requests so that I can contribute the updated version of pyfingrid.py back to the community + de-coupled sample scripts (one for the the authentication part and another for fetching the data) :-P
